### PR TITLE
Glimmer.js HN PWA

### PR DESCRIPTION
**Pull request by mhadaily**
Glimmer.js HN PWA, Please see #83

### CODE REVIEW

**Comment by addyosmani** [comment]
Wooo. Glimmer! Thanks for the PR, @mhadaily! 🎉 

I'll take a deeper look at performance and the traces later in the week. Took a quick review now:

- UX looks great. I love the skeleton screens you adopted for content placeholders.
- Application Shell pattern and offline asset caching appears to work as expected
- Navigating through the app feels pretty smooth

One routing bug I ran into was if you navigate through paginated pages the route doesn't get updated (meaning that a linking to the current route doesn't return back the same content):

image

The other apps currently respect routing for these pages:

https://hn.kristoferbaxter.com/top/3
https://react-hn.kristoferbaxter.com/top/3
https://vue-hn.now.sh/top/3

**Comment by mhadaily** [comment]
@addyosmani I could produce the bug that you mentioned and I have fixed that. the app now respects routing with page number in hash. 
https://glimmer-hn-pwa.firebaseapp.com/#/news/1
https://glimmer-hn-pwa.firebaseapp.com/#/newest/11
https://glimmer-hn-pwa.firebaseapp.com/#/news/10
and etc.
The webpage result is pretty much similar so I haven't update the links.

**Comment by addyosmani** [comment]
Friendly ping on that one routing issue :)

**Comment by mhadaily** [comment]
Hi
I implemented a quick fix last week, I thought I sent a message. Please
check it out now thanks.
Sorry for not following up, I'm on a trip...

Thanks

File: site/_apps/glimmer-hn-pwa.txt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+api: 'Node-hnapi (unofficial)'
+hosting: 'Firebase'
+authors:
+    - name: 'Majid'
+lighthouse: '91/100'
+interactive-em: '3.4s'
+interactive-faster-3g: '2.4s'
+lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170612_BV_5cd2c04584152647e2e556f5af8f5eba&run=3
+wpt-em-link: https://www.webpagetest.org/result/170708_DK_bcd8d8c3da33a5f7badc29417f3b95c7/
+wpt-faster-3g-link: https://www.webpagetest.org/result/170708_34_1ae3fbe0c7f0f90ad0b46282ab2ce5a5/
```
Comment (addyosmani) [inline site/_apps/glimmer-hn-pwa.txt]: New run through WPT for faster 3G:
https://www.webpagetest.org/result/170719_6Y_4ebdb9a788a0691f0c5159e40286545b/
interactive-faster-3g: 5.120s

Lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170719_6Y_4ebdb9a788a0691f0c5159e40286545b&run=2

File: site/_apps/glimmer-hn-pwa.txt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+server-side-rendering: 'None'
+api: 'Node-hnapi (unofficial)'
+hosting: 'Firebase'
+authors:
+    - name: 'Majid'
+lighthouse: '91/100'
+interactive-em: '3.4s'
+interactive-faster-3g: '2.4s'
+lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170612_BV_5cd2c04584152647e2e556f5af8f5eba&run=3
+wpt-em-link: https://www.webpagetest.org/result/170708_DK_bcd8d8c3da33a5f7badc29417f3b95c7/
```
Comment (addyosmani) [inline site/_apps/glimmer-hn-pwa.txt]: New wpt-em-link: https://www.webpagetest.org/result/170720_F4_fe90cf68506962341fd78fda3ef9dc63/

interactive-em: 6.250s

**Comment by addyosmani** [comment]
URLs I've checked:

https://glimmer-hn-pwa.firebaseapp.com/#/news/1
https://glimmer-hn-pwa.firebaseapp.com/#/item/14806986
https://glimmer-hn-pwa.firebaseapp.com/#/newest/4
https://glimmer-hn-pwa.firebaseapp.com/#/show/1
https://glimmer-hn-pwa.firebaseapp.com/#/item/14805750
https://glimmer-hn-pwa.firebaseapp.com/#/ask/3
https://glimmer-hn-pwa.firebaseapp.com/#/jobs/1

Routes appear to be solid.

responsive + pagination UX, font-size are on point.

I just did a new run through WebPageTest with the latest deployed version and linked up the reports above. It's looking like the numbers are a little closer to 5s and 6s for 3G and EM time to interactive now (a little slower than before). 

@anubhav7495 Would you be interested in an audit of this one?

**Comment by anubhav7495** [comment]
@mhadaily Nice app man! Super cool box shadow hover effect in posts :sunglasses:

#### Some bugs I could find in the app: 

1. `Ask HN` links when clicked open a new tab with `top` view.
2. **Minor**: collapse links `[-]` in comments view do nothing when clicked.

#### Performance:

[LightHouse Faster_3G](https://www.webpagetest.org/lighthouse.php?test=170720_FY_7fe71744797e12811385cc4c3bf6ed13&run=1) TTI `4.8s`

#### Suggestions:

1. Try `dns-prefetch` for `node-hnapi` endpoint, you would save `300ms` on TTI on average.
2. [WPT_waterfall](https://www.webpagetest.org/result/170720_FY_7fe71744797e12811385cc4c3bf6ed13/1/details/#waterfall_view_step1) Here, the network request is the last thing that happens, maybe try to fire it a little sooner.
3. In the waterfall above, `Request 13` is `lighthouse.html`, I don't think you need that for the app.

I would not list them all but there are many things you can optimize here to get your TTI Lower.

Thanks.

**Comment by addyosmani** [comment]
Thanks for the review, @anubhav7495!

**Comment by mhadaily** [comment]
@addyosmani and @anubhav7495 Thanks for the feedback. I have spent some time and fixed the bugs as well as optimized more. Please see new results 

From: Dulles, VA - Moto G4 - Chrome Beta - 3G_EM
7/24/2017, 10:24:35 PM
https://www.webpagetest.org/result/170724_59_43187e264846ff2314364b01642ee1e1/
Fully Loaded: 4.037s

Only 10 requests in [Waterfall](https://www.webpagetest.org/result/170724_59_43187e264846ff2314364b01642ee1e1/1/details/#waterfall_view_step1): 
----------------------------------------------------------
From: Dulles, VA - Moto G4 - Chrome Beta - 3G
7/24/2017, 10:29:20 PM
https://www.webpagetest.org/result/170724_Q5_ae999c76745dd61df39f744a6b10f45b/
Fully loaded: 3.100s
Only 10 requests in [Waterfall](https://www.webpagetest.org/result/170724_Q5_ae999c76745dd61df39f744a6b10f45b/1/details/#waterfall_view_step1): 

------
Lighthouse performance: https://www.webpagetest.org/lighthouse.php?test=170724_Q5_ae999c76745dd61df39f744a6b10f45b&run=2

thank you for your valuable feedbacks and review.

**Comment by anubhav7495** [comment]
@mhadaily Nice work man!

I still see 2 issues

1. Ask HN Links in `Top` or `New` views still, have the same behavior as before.
2. Lighthouse scores do not match.
**Faster 3G**: 4.6s
**3G Emerging Markets**: 5.1s

We use the `First Interactive` or `Consistently Interactive` from Linked Lighthouse report for determining the `TTI`.

Thanks.

**Comment by mhadaily** [comment]
@anubhav7495 number 1 has been fixed and I tried to optimize even more and now I got the new results:
1- 3G_EM
* https://www.webpagetest.org/result/170731_H3_07fad36e6a920575287dbd0b02bc1288/
* https://www.webpagetest.org/lighthouse.php?test=170731_H3_07fad36e6a920575287dbd0b02bc1288&run=2

2- 3G
* https://www.webpagetest.org/result/170731_DC_b92d9dc97fb6acc21996117a817e7621/
* https://www.webpagetest.org/lighthouse.php?test=170731_DC_b92d9dc97fb6acc21996117a817e7621&run=3

as you see I could manage to reduce to **3.6 in 3G_EM** and **2.8 in 3G.** 

What I didn't understand is that "First Interactive" for both 3G and 3G_EM must be the same? well, "Lighthouse scores do not match." for this, I tried many times, it never matches. could you please clarify? or if now it's fine please let me know. 

thank you

**Comment by anubhav7495** [comment]
@mhadaily Nice work man!

By `Lighthouse scores do not match` we mean that `TTI` metric from the linked lighthouse reports are not equal to ones specified in your pull request.

You need to change the `interactive-em` and `interactive-3g` times from `3.4s` & `2.4s` respectively to `3.6s ` & `2.8s` in your app description `txt` file.

Other than that, i see no more issues from my side. Awesome implementation man!

@addyosmani over to you.

File: site/_apps/glimmer-hn-pwa.txt [hunk 3]
```diff
@@ -14,11 +14,11 @@ hosting: 'Firebase'
...
     - name: 'Majid'
 lighthouse: '91/100'
-interactive-em: '3.4s'
-interactive-faster-3g: '2.4s'
-lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170612_BV_5cd2c04584152647e2e556f5af8f5eba&run=3
-wpt-em-link: https://www.webpagetest.org/result/170708_DK_bcd8d8c3da33a5f7badc29417f3b95c7/
-wpt-faster-3g-link: https://www.webpagetest.org/result/170708_34_1ae3fbe0c7f0f90ad0b46282ab2ce5a5/
+interactive-em: '3.6s'
+interactive-faster-3g: '2.8s'
+lighthouse-link: https://www.webpagetest.org/lighthouse.php?test=170731_DC_b92d9dc97fb6acc21996117a817e7621&run=3
```
Comment (addyosmani) [inline site/_apps/glimmer-hn-pwa.txt]: Confirmed this are now correct per @anubhav7495's last request on the review. Thanks for updating them!

**Comment by addyosmani** [comment]
These traces lgtm based on our requirements. Landing. Thanks for working on feedback, @mhadaily and for conducting a second pass @anubhav7495. Feel free to tweet about the implementation and we'll RT.

**Pull request merged**